### PR TITLE
[libhail] SArray and SArrayValue

### PR DESCRIPTION
**Pull request by johnc1231**
This PR introduces `SArray` and `SArrayValue`, lets us compile with arrays. I added implementations for `MakeArray` and `ArrayRef`. 

To help myself out, I also registered the region allocation function on the `CompileModule` so it's always available to call. I added a `printf` function as well that takes a C++ string and a vector of `llvm::Value` to allow arbitrary debugging prints.

### CODE REVIEW

File: libhail/src/hail/query/backend/stype.cpp [hunk 1]
```diff
@@ -63,6 +63,12 @@ SFloat32::SFloat32(const Type *type)
 SFloat64::SFloat64(const Type *type)
   : SType(self_tag, type) {}
 
+SArray::SArray(const Type *type, const SType *elementType)
```
Comment (cseed) [inline libhail/src/hail/query/backend/stype.cpp:66]: `element_type`, variables use snake case.

File: libhail/src/hail/query/backend/stype.cpp [hunk 2]
```diff
@@ -63,6 +63,12 @@ SFloat32::SFloat32(const Type *type)
...
+  : SType(self_tag, type), elementType(elementType) {}
+
+SCanonicalArray::SCanonicalArray(const Type *type, const SType *elementType)
```
Comment (cseed) [inline libhail/src/hail/query/backend/stype.cpp:69]: Same.

File: libhail/src/hail/query/backend/stype.hpp [hunk 3]
```diff
@@ -77,6 +77,18 @@ class SFloat64 : public SType {
   SFloat64(const Type *type);
 };
 
+class SArray : public SType {
+  public:
+    static const Tag self_tag = SType::Tag::ARRAY;
```
Comment (cseed) [inline libhail/src/hail/query/backend/stype.hpp:82]: This won't work as it stands.  Only concrete members of the hierarchy can have tags, so we shouldn't have this, and there should be a `Tag::CANONICAL_ARRAY`.

There is a question if we should have SArray at all.  If you want to be able to do `isa<SArray>(stype)` or `cast<SArray>(stype)`, then we enhance the casting stuff to support non-concrete intermediate classes (I would probably avoid concrete intermediate classes for now).

File: libhail/src/hail/query/backend/svalue.cpp [hunk 4]
```diff
@@ -46,6 +46,20 @@ SFloat32Value::SFloat32Value(const SType *stype, llvm::Value *value)
 SFloat64Value::SFloat64Value(const SType *stype, llvm::Value *value)
   : SValue(self_tag, stype), value(value) {}
 
+SCanonicalArrayValue::SCanonicalArrayValue(const SArray *stype, llvm::Value *length, llvm::Value *missing, llvm::Value *data)
```
Comment (cseed) [inline libhail/src/hail/query/backend/svalue.cpp:49]: `stype` should be a `SCanonicalArray`, no?

File: libhail/src/hail/query/backend/svalue.cpp [hunk 5]
```diff
@@ -46,6 +46,20 @@ SFloat32Value::SFloat32Value(const SType *stype, llvm::Value *value)
...
+  : SArrayValue(stype), length(length) {
+}
+
+SCanonicalArrayValue::~SCanonicalArrayValue() {
+  // TODO: Implement?
+}
+
+SInt64Value*
+SCanonicalArrayValue::getLength(TypeContext *tc) {
```
Comment (cseed) [inline libhail/src/hail/query/backend/svalue.cpp:58]: Members are snake case: `get_length`.

File: libhail/src/hail/query/backend/svalue.hpp [hunk 6]
```diff
@@ -63,6 +64,25 @@ class SFloat64Value : public SValue {
   SFloat64Value(const SType *stype, llvm::Value *value);
 };
 
+class SArrayValue : public SValue {
+public:
+  static const Tag self_tag = SValue::Tag::ARRAY;
+  const SArray *const stype;
+  SArrayValue(const SArray *stype): SValue(self_tag, stype), stype(stype) {};
+  virtual SValue getElement(SInt64Value idx)=0;
+  virtual SInt64 getLength()=0;
```
Comment (cseed) [inline libhail/src/hail/query/backend/svalue.hpp:73]: Spaces: `...() = 0`.

File: libhail/src/hail/query/backend/svalue.hpp [hunk 7]
```diff
@@ -63,6 +64,25 @@ class SFloat64Value : public SValue {
...
+ };
+
+class SCanonicalArrayValue : public SArrayValue {
+  public:
+    SCanonicalArrayValue(const SArray *stype, llvm::Value *length, llvm::Value *missing, llvm::Value *data);
+    SInt64Value* getLength(TypeContext *tc);
+    ~SCanonicalArrayValue();
+  private:
+    llvm::Value *data;
+    llvm::Value *length;
```
Comment (cseed) [inline libhail/src/hail/query/backend/svalue.hpp:83]: Don't  you need the missing pointer here, too?

File: libhail/src/hail/query/backend/svalue.hpp [hunk 8]
```diff
@@ -63,6 +64,25 @@ class SFloat64Value : public SValue {
   SFloat64Value(const SType *stype, llvm::Value *value);
 };
 
+class SArrayValue : public SValue {
+public:
+  static const Tag self_tag = SValue::Tag::ARRAY;
```
Comment (cseed) [inline libhail/src/hail/query/backend/svalue.hpp:69]: Same comments as SArray.

**Review by cseed — CHANGES_REQUESTED** [review]

File: libhail/src/hail/query/backend/compile.cpp [hunk 9]
```diff
@@ -218,6 +218,13 @@ CompileFunction::emit(IsNA *x) {
   return EmitValue(m, new SBoolValue(new SBool(ir_type(x)), v));
 }
 
+EmitValue
+CompileFunction::emit(ArrayLen *x) {
+  auto arrayDataValue = emit(x->get_child(0)).as_data(*this);
+  auto arraySValue = cast<SArrayValue>(arrayDataValue.svalue);
```
Comment (cseed) [inline libhail/src/hail/query/backend/compile.cpp:224]: Variables are snake_case.

File: libhail/src/hail/query/backend/compile.cpp [hunk 10]
```diff
@@ -218,6 +218,13 @@ CompileFunction::emit(IsNA *x) {
...
+  return EmitValue(arrayDataValue.missing, arraySValue->get_length(&this->tc));
```
Comment (cseed) [inline libhail/src/hail/query/backend/compile.cpp:225]: You don't need to use `this->`.

File: libhail/src/hail/query/backend/stype.hpp [hunk 11]
```diff
@@ -77,6 +79,15 @@ class SFloat64 : public SType {
   SFloat64(const Type *type);
 };
 
+class SCanonicalArray : public SType{
```
Comment (cseed) [inline libhail/src/hail/query/backend/stype.hpp:82]: Space before `{`.

File: libhail/src/hail/query/backend/stype.hpp [hunk 12]
```diff
@@ -77,6 +79,15 @@ class SFloat64 : public SType {
...
+  public:
```
Comment (cseed) [inline libhail/src/hail/query/backend/stype.hpp:83]: Access specifiers are flush with the class definition and indent is 2.  So:

```
class SCanonicalArray :  public SType {
public:
  static const Tag ...
}
```

File: libhail/src/hail/query/backend/svalue.cpp [hunk 13]
```diff
@@ -2,6 +2,7 @@
 #include <hail/query/backend/svalue.hpp>
 #include <llvm/IR/BasicBlock.h>
 #include <llvm/IR/IRBuilder.h>
+#include <stdint.h>
```
Comment (cseed) [inline libhail/src/hail/query/backend/svalue.cpp:5]: Prefer C++ headers: `#include <cstdint>`.

File: libhail/src/hail/query/backend/svalue.cpp [hunk 14]
```diff
@@ -46,6 +47,28 @@ SFloat32Value::SFloat32Value(const SType *stype, llvm::Value *value)
 SFloat64Value::SFloat64Value(const SType *stype, llvm::Value *value)
   : SValue(self_tag, stype), value(value) {}
 
+SCanonicalArrayValue::SCanonicalArrayValue(const SCanonicalArray *stype, llvm::Value *length, llvm::Value *missing, llvm::Value *data)
+  : SArrayValue(stype), length(length), missing(missing), data(data) {
+}
+
+SCanonicalArrayValue::~SCanonicalArrayValue() {
+  // TODO: Implement?
```
Comment (cseed) [inline libhail/src/hail/query/backend/svalue.cpp:55]: Nothing needed.

File: libhail/src/hail/query/backend/svalue.cpp [hunk 15]
```diff
@@ -46,6 +47,28 @@ SFloat32Value::SFloat32Value(const SType *stype, llvm::Value *value)
...
+}
+
+SInt64Value*
```
Comment (cseed) [inline libhail/src/hail/query/backend/svalue.cpp:58]: Space before *.

File: libhail/src/hail/query/backend/svalue.cpp [hunk 16]
```diff
@@ -46,6 +47,28 @@ SFloat32Value::SFloat32Value(const SType *stype, llvm::Value *value)
...
+SCanonicalArrayValue::get_length(TypeContext *tc) const {
```
Comment (cseed) [inline libhail/src/hail/query/backend/svalue.cpp:59]: Use reference when the pointer can't be null.

File: libhail/src/hail/query/backend/svalue.cpp [hunk 17]
```diff
@@ -46,6 +47,28 @@ SFloat32Value::SFloat32Value(const SType *stype, llvm::Value *value)
...
+  auto returnType = new SInt64(tc->tint32);
```
Comment (cseed) [inline libhail/src/hail/query/backend/svalue.cpp:60]: `return_type`.

File: libhail/src/hail/query/backend/svalue.cpp [hunk 18]
```diff
@@ -46,6 +47,28 @@ SFloat32Value::SFloat32Value(const SType *stype, llvm::Value *value)
...
+  return new SInt64Value(returnType, length);
+}
+
+SValue*
+SCanonicalArrayValue::get_element(CompileFunction &cf, SInt64Value *idx) {
+  // TODO check for missing.
+  auto stride_value = llvm::ConstantInt::get(llvm::Type::getInt64Ty(cf.llvm_context), static_cast<uint64_t>(this->stype->element_stride));
```
Comment (cseed) [inline libhail/src/hail/query/backend/svalue.cpp:67]: I don't think you need the static cast here.  Also, no, `this->`.

File: libhail/src/hail/query/backend/svalue.hpp [hunk 19]
```diff
@@ -63,6 +64,27 @@ class SFloat64Value : public SValue {
   SFloat64Value(const SType *stype, llvm::Value *value);
 };
 
+class SArrayValue : public SValue {
+public:
+  static const Tag self_tag = SValue::Tag::CANONICALARRAY;
+  const SCanonicalArray *const stype;
+  SArrayValue(const SCanonicalArray *stype): SValue(self_tag, stype), stype(stype) {};
```
Comment (cseed) [inline libhail/src/hail/query/backend/svalue.hpp:71]: No semi after function definition.

File: libhail/src/hail/query/backend/svalue.hpp [hunk 20]
```diff
@@ -63,6 +64,27 @@ class SFloat64Value : public SValue {
...
+  virtual SValue* get_element(CompileFunction &cf, SInt64Value *idx) = 0;
+  virtual SInt64Value* get_length(TypeContext *tc) const = 0;
```
Comment (cseed) [inline libhail/src/hail/query/backend/svalue.hpp:73]: `type *function()` not `type* function()`.

File: libhail/src/hail/query/ir_type.cpp [hunk 21]
```diff
@@ -43,12 +43,24 @@ IRType::infer(IsNA *x) {
...
 
+const Type *
+IRType::infer(ArrayLen *x) {
+  infer(x->get_child(0));
+  return tc.tint64;
+}
+
+const Type *
+IRType::infer(ArrayRef *x) {
+  auto inferedChild = infer(x->get_child(0));
```
Comment (cseed) [inline libhail/src/hail/query/ir_type.cpp:54]: `infered_child` (maybe `child_type` is better?)

File: libhail/src/hail/query/backend/svalue.cpp [hunk 22]
```diff
@@ -46,6 +47,28 @@ SFloat32Value::SFloat32Value(const SType *stype, llvm::Value *value)
...
+  auto element_addr = cf.llvm_ir_builder.CreateGEP(this->data, cf.llvm_ir_builder.CreateMul(idx->value, stride_value));
+  return this->stype->element_type->load_from_address(cf, element_addr);
```
Comment (cseed) [inline libhail/src/hail/query/backend/svalue.cpp:69]: No `this->` needed.

**Review by cseed — CHANGES_REQUESTED** [review]
Mostly formatting, naming, etc.  Code itself looks good.

**Comment by johnc1231** [comment]
Bump @cseed

**Comment by johnc1231** [comment]
Bump @cseed

**Comment by danking** [comment]
I'm closing this because its so old. We can start a new PR when we are ready to revisit this.

**Pull request closed without merging**